### PR TITLE
MISP collector and parser bots

### PR DESCRIPTION
**Pull request by kralca**

### CODE REVIEW

File: intelmq/bots/collectors/misp/collector.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+from intelmq.lib.bot import Bot
+from intelmq.lib.harmonization import DateTime
+from intelmq.lib.message import Report
+
+
+class MISPCollectorBot(Bot):
+
+    def init(self):
+        self.logger.info('MISPCollectorBot initialising')
```
Comment (sebix) [inline intelmq/bots/collectors/misp/collector.py]: This is already logged by the botclass itself providing more information.

File: intelmq/bots/collectors/misp/collector.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    'Content-Type': 'application/json',
+                })
+                tag = self.parameters.misp_tag_to_process
+                post_data = {
+                    'request': {
+                        'Event': {
+                            'tag': tag,
+                            'id': misp_event_id,
+                }}}
+                session.post(self.misp_del_tag_url, data=json.dumps(post_data))
```
Comment (sebix) [inline intelmq/bots/collectors/misp/collector.py]: I'm not sure if this should be done after sending the messages (when they are "safe")?

File: intelmq/bots/BOTS [hunk 3]
```diff
@@ -421,6 +421,18 @@
                 "rate_limit": 180000
             }
         },
+        "MISP Generic": {
+            "description": "Collect events from a MISP server.",
+            "module": "intelmq.bots.collectors.misp.collector",
+            "parameters": {
+                "feed": "misp_generic",
+                "misp_url": "<insert url of MISP server (with trailing '/')>",
```
Comment (sebix) [inline intelmq/bots/BOTS]: The alphabetic order is: misp_key, misp_tag_processed, misp_tag_to_process, misp_url

File: intelmq/bots/parsers/misp/parser.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+        'ip-dst': 'source.ip',
+        'email-src': 'source.account',
+        'url': 'source.url',
+    }
+
+    def process(self):
+        self.logger.info('MISPParserBot processing message from collector')
+
+        report = self.receive_message()
+        if not report or not report.contains('raw'):
```
Comment (sebix) [inline intelmq/bots/parsers/misp/parser.py]: This block is not necessary anymore.

File: intelmq/bots/parsers/misp/parser.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        'md5': 'malware.hash.md5',
+        'sha1': 'malware.hash.sha1',
+        'ip-src': 'source.ip',
+        'ip-dst': 'source.ip',
+        'email-src': 'source.account',
+        'url': 'source.url',
+    }
+
+    def process(self):
+        self.logger.info('MISPParserBot processing message from collector')
```
Comment (sebix) [inline intelmq/bots/parsers/misp/parser.py]: Also done by Bot-class

**Comment by sebix** [comment]
Code looks fine.

I'm missing tests for the parser.

**Comment by kralca** [comment]
Please check if latest commits address feedback. Thanks sebix!

File: intelmq/tests/bots/parsers/misp/test_parser.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+        """ Test if correct Event has been produced. """
+        self.run_bot()
+        self.assertMessageEqual(0, EXAMPLE_EVENT)
+
+    def assertMessageEqual(self, queue_pos, expected_msg):
+        """
+        Override test.BotTestCase.assertMessageEqual to allow
+        comparison of the 'raw' as a dict. The MISP event is a JSON
+        encoded dict which serializes non-deterministically because a
+        dict is not ordered.
```
Comment (sebix) [inline intelmq/tests/bots/parsers/misp/test_parser.py]: You just use `json.dumps(obj, sort_keys=True)` to sort the object's keys.

File: intelmq/bots/BOTS [hunk 7]
```diff
@@ -421,6 +421,18 @@
                 "rate_limit": 180000
             }
         },
+        "MISP Generic": {
```
Comment (sebix) [inline intelmq/bots/BOTS]: 'MI' < 'Ma'

Sorry, the sorting is not lexical.

**Comment by sebix** [comment]
Merged along with the fix for the BOTS-file. Thanks for your contribution!

**Pull request merged**